File: AoC_2021/day_4/code4.py
# star 1
with open('data.txt') as f:
    data = f.readlines()
data = [d.strip() for d in data if d.strip() != '']

nrs = [int(d) for d in data[0].split(",")]

data_split = [d.replace("  ", " ").split(" ") for d in data[1:]]
flattened = [int(el) for row in data_split for el in row]

def nr_to_none(full_list, nr):
    return [None if f == nr else f for f in full_list]

def check_bingo_rows(flattened):
    start_inds = range(0, len(flattened) - 5, 5)
    return([si for si in start_inds if not any(flattened[si:si+5])])
   
def check_bingo_cols(flattened):
    start_inds = [r for r in range(0, len(flattened)) if r % 25 < 5]
    def subset_col_from_si(si):
        return [flattened[i] for i in range(si, si+21, 5)]
    return([si for si in start_inds if not any(subset_col_from_si(si))])

bingo = False
while not bingo:
    flattened = nr_to_none(flattened, nrs[0])
    si_bingo = check_bingo_rows(flattened) + check_bingo_cols(flattened)
    if len(si_bingo) > 0:
        break
    nrs.pop(0)

bingo_card_won = [flattened[i] for i in range(si_bingo[0], si_bingo[0] + 25)]
card_won_nrs_remaining = [nr for nr in bingo_card_won if nr is not None]

print(nrs[0] * sum(card_won_nrs_remaining))

File: AoC_2021/day_4/code4.R
# star 1

data <- readLines("day_4/data.txt")
nrs  <- strsplit(data[1], ",") |> unlist() |> as.integer()
cards_raw <- data[-c(1:2)][data[-c(1:2)] != ""]
cards_clean <- gsub("  ", " ", cards_raw) |> strsplit(" ") |>
  lapply(function(x) x[x!=""] |> as.integer())
cards <- split(cards_clean, rep(1:(length(cards_clean) / 5), each = 5)) |>
  lapply(function(x) unlist(x) |> matrix(byrow = TRUE, nrow = 5))

cross_nr_card <- function(card, nr) {
  card[card == nr] <- NA
  card
}

check_bingo <- function(card) {
  if (any(colSums(is.na(card)) == 5) | any(rowSums(is.na(card)) == 5))
    return(card)
}

star_1 <- function(data, nrs){
  while (TRUE) {
    cards <- lapply(cards, cross_nr_card, nrs[1])
    bingo <- lapply(cards, check_bingo)
    if (!all(sapply(bingo, is.null))) {
      break
    }
    nrs <- nrs[-1] 
  }
  unlist(bingo[!sapply(bingo, is.null)]) |> sum(na.rm = TRUE) * nrs[1]
}

star_1(data, nrs)

# star 2
star_2 <- function(data, nrs) {
  while (TRUE) {
    last_bingo <- lapply(cards, check_bingo)
    cards <- lapply(cards, cross_nr_card, nrs[1])
    bingo <- lapply(cards, check_bingo)
    if (sum(sapply(bingo, is.null)) == 0) {
      break
    }
    nrs <- nrs[-1] 
  }
  cards[[which(sapply(last_bingo, is.null))]] |> sum(na.rm = TRUE) * nrs[1]
}

star_2(data, nrs)